src/NServiceBus.Core/SettingsExtensions.cs

```diff
@@ -20,16 +20,19 @@ public static T GetConfigSection<T>(this ReadOnlySettings settings) where T : cl
             var typesToScan = settings.GetAvailableTypes();
             var configurationSource = settings.Get<IConfigurationSource>();
 
-            // ReSharper disable HeapView.SlowDelegateCreation
-            var sectionOverrideType = typesToScan.Where(t => !t.IsAbstract)
-                .FirstOrDefault(t => typeof(IProvideConfiguration<T>).IsAssignableFrom(t));
-            // ReSharper restore HeapView.SlowDelegateCreation
+            var sectionOverrideTypes = typesToScan.Where(t => !t.IsAbstract && typeof(IProvideConfiguration<T>).IsAssignableFrom(t)).ToArray();
+            if (sectionOverrideTypes.Length > 1)
+            {
+                throw new Exception($"{sectionOverrideTypes.Length} configuration providers implementing IProvideConfiguration<{typeof(T).Name}> were found. Ensure that only one configuration provider per configuration section is found to resolve this error.");
```

Comment:
The idea of listing the types is a decent suggestion as well, since we already have the type names right there.